numpy/core/src/multiarray/nditer_constr.c

```diff
@@ -2587,16 +2578,13 @@ npyiter_new_temp_array(NpyIter *iter, PyTypeObject *subtype,
             else {
                 if (shape == NULL) {
                     /*
-                     * If deleting this axis produces a reduction, but
-                     * reduction wasn't enabled, throw an error.
-                     * NOTE: We currently always allow new-axis if the iteration
-                     *       size is 1 (thus allowing broadcasting sometimes).
+                     * Even the shape == NULL (which happens for allocated
+                     * arrays), is already checked in `npyiter_fill_axisdata`.
                      */
                     if (!reduction_axis && NAD_SHAPE(axisdata) != 1) {
-                        if (!npyiter_check_reduce_ok_and_set_flags(
-                                iter, flags, op_itflags, i)) {
-                            return NULL;
-                        }
+                        /* We do not know which op this is so skip mask test */
+                        assert(npyiter_check_reduce_ok_and_set_flags(
+                               iter, flags, op_itflags, 0, -1, i));
```

Comment:
I had not removed this before, because it seems necessary, and the general path is of course hit.  But. If `shape==NULL`, this means that `op==NULL` (otherwise `shape==PyArray_DIMS(op)`).  If `op_axes` are given however, and `op_axes` includes `-1` then the last `npyiter_check_reduce_ok_and_set_flags` above is already hit.

I.e. the first gut-feeling that this function (allocating the array) is not in charge of checking this, is actually correct.

Reply:
This makes sense, but doesn't that mean that the whole clause can be removed? (Or is that what you were referring to with a clean-up follow-up?) I.e. would `assert(shape != NULL)` already work in this `else` clause?

Reply:
Hmmm, I was thinking this might go away, but had not planned concretely. It was more an abundance of caution thing... I think I will just delete it with a note that `-1` is not relevant for *allocating* arrays, which this function is after all about. (shape may or may not be `NULL` here, so not sure what I could assert for that).
The tests should be pretty good about checking the code paths that could in theory run into the assert if the previous part was buggy.

Reply:
OK, makes sense.

Reply:
My first idea was just to make do some of the flag checking earlier to make a new itflag (not to check 3 different flags for reduction).  That may not be bad, but my attempts to redesign the broadcast logic a bit have not yet fruited. Likely I will just give up for the time being...